vault/vault.go

```diff
@@ -2,192 +2,238 @@ package vault
-		// Vault Token
-		token = getVaultParam(secretConfig, VaultTokenKey)
-		if token == "" {
-			return nil, ErrVaultTokenNotSet
-		}
-		// Vault Address
-		address := getVaultParam(secretConfig, api.EnvVaultAddress)
-		if address == "" {
-			return nil, ErrVaultAddressNotSet
-		}
 
-		config.Address = address
-		// Get TLS Settings
-		tlsConfig := api.TLSConfig{}
-		skipVerify := getVaultParam(secretConfig, api.EnvVaultInsecure)
-		if skipVerify != "" {
-			insecure, err := strconv.ParseBool(skipVerify)
-			if err != nil {
-				return nil, ErrInvalidSkipVerify
-			}
-			tlsConfig.Insecure = insecure
-		}
-		// Get Cert Paths
-		cacert := getVaultParam(secretConfig, api.EnvVaultCACert)
-		tlsConfig.CACert = cacert
-		capath := getVaultParam(secretConfig, api.EnvVaultCAPath)
-		tlsConfig.CAPath = capath
-		clientcert := getVaultParam(secretConfig, api.EnvVaultClientCert)
-		tlsConfig.ClientCert = clientcert
-		clientkey := getVaultParam(secretConfig, api.EnvVaultClientKey)
-		tlsConfig.ClientKey = clientkey
-		tlsserverName := getVaultParam(secretConfig, api.EnvVaultTLSServerName)
-		tlsConfig.TLSServerName = tlsserverName
-		err := config.ConfigureTLS(&tlsConfig)
-		if err != nil {
-			return nil, err
-		}
+	if len(secretConfig) == 0 && config.Error != nil {
+		return nil, config.Error
 	}
-	// Vault craps out if address is not in correct format
-	if !strings.HasPrefix(config.Address, vaultAddressPrefix) {
+
+	token := getVaultParam(secretConfig, api.EnvVaultToken)
+	if token == "" {
+		return nil, ErrVaultTokenNotSet
+	}
+
+	address := getVaultParam(secretConfig, api.EnvVaultAddress)
+	if address == "" {
+		return nil, ErrVaultAddressNotSet
+	}
+	// Vault fails if address is not in correct format
+	if !strings.HasPrefix(address, vaultAddressPrefix) {
 		return nil, ErrInvalidVaultAddress
 	}
+	config.Address = address
+
+	if err := configureTLS(config, secretConfig); err != nil {
+		return nil, err
+	}
 
-	client, err := getVaultClient(config)
+	client, err := newVaultClient(config)
 	if err != nil {
 		return nil, err
 	}
+	client.SetToken(token)
 
-	if token != "" {
-		// Set the token if present
-		client.SetToken(token)
+	backendPath := getVaultParam(secretConfig, VaultBackendPathKey)
+	if backendPath != "" {
+		if !strings.HasSuffix(backendPath, "/") {
+			backendPath = backendPath + "/"
+		}
+	} else {
+		backendPath = DefaultBackendPath
 	}
+
 	return &vaultSecrets{
-		endpoint: config.Address,
-		client:   client,
+		endpoint:    config.Address,
+		client:      client,
+		backendPath: backendPath,
 	}, nil
-
 }
 
 func (v *vaultSecrets) String() string {
 	return Name
 }
 
+func (v *vaultSecrets) getSecretKey(secretID string, isKvV2 bool) string {
+	if isKvV2 {
+		return v.backendPath + kvDataKey + "/" + secretID
+	}
+	return v.backendPath + secretID
+}
+
+func (v *vaultSecrets) isKvBackendV2() (bool, error) {
+	mounts, err := v.client.Sys().ListMounts()
+	if err != nil {
+		return false, err
+	}
+
+	for path, mount := range mounts {
```

Comment:
Why not find out the backend version only once while creating the client and then use it in each API?